### PR TITLE
Refactor Monitoring and Server Selector Unit Tests

### PR DESCRIPTION
**Pull request by GabrielEValenzuela**
|Related issue|
|---|
|#25415|

### Description:
This pull request addresses the refactoring of the `monitoring_test.cpp` and `serverSelector_test.cpp` files to improve the efficiency of mocking the `HTTPRequest` class, avoiding unnecessary testing of external connections and out-of-scope classes or libraries. The changes also include the injection of a mock `Monitor` class in the server selector tests, enabling more granular and accurate testing of the class conditions.

Additionally, the outdated `fakeOpenSearchServer.hpp` file has been removed as it is no longer needed following these test improvements.

**Changes Include:**
- Reimplemented `monitoring_test.cpp` with enhanced mocking of `HTTPRequest`.
- Refactored `serverSelector_test.cpp` to inject a mock `Monitor` class.
- Removed `fakeOpenSearchServer.hpp` file.

> [!NOTE]
> While component tests are beyond the scope of this issue, a minor refactor related to dead time handling was necessary to ensure the tests were functional and reliable.

### CODE REVIEW

File: src/engine/source/indexerconnector/include/indexerConnector.hpp [hunk 1]
```diff
@@ -22,9 +22,10 @@
 #define EXPORTED
 #endif
 
-static constexpr auto DEFAULT_INTERVAL = 60u;
+static constexpr auto DEFAULT_INTERVAL = 60000u;
```
Comment (sebasfalcone) [inline src/engine/source/indexerconnector/include/indexerConnector.hpp:25]: Question: Why did you change the interval?

File: src/engine/source/indexerconnector/src/monitoring.hpp [hunk 2]
```diff
@@ -62,9 +62,14 @@ enum Type : std::size_t
 
 /**
  * @brief Monitoring class.
+ * This class is used to monitor the health of the servers.
```
Comment (sebasfalcone) [inline src/engine/source/indexerconnector/src/monitoring.hpp:65]: Nitpick: Doxygen supports notes. There are appended below the brief
```suggestion
 *
 * @note This class is used to monitor the health of the servers.
```

File: src/engine/source/indexerconnector/test/unit/serverSelector_test.cpp [hunk 3]
```diff
@@ -162,21 +180,59 @@ TEST_F(ServerSelectorTest, TestGetNextBeforeAndAfterHealthCheck)
...
+                               "manager\tshards\tpri\trelo\tinit\tunassign\tpending_tasks\tmax_task_wait_time\tactive_"
+                               "shards_percent\n1725296432\t17:00:32\twazuh-"
+                               "cluster\tred\t1\t1\ttrue\t47\t47\t0\t0\t0\t0\t-\t100.0%\n";
+                    postRequestParameters.onError(response.c_str(), 200);
+                }
+                else
+                {
+                    // Unknown server, simulate an error
+                    postRequestParameters.onError("Unknown server", 404);
+                    return;
```
Comment (sebasfalcone) [inline src/engine/source/indexerconnector/test/unit/serverSelector_test.cpp:219]: Improve: Missing yellow server

**Review by sebasfalcone — CHANGES_REQUESTED** [review]

File: src/engine/source/indexerconnector/include/indexerConnector.hpp (hunk 1 above)
Comment (GabrielEValenzuela) [reply src/engine/source/indexerconnector/include/indexerConnector.hpp:25]: Time scale. We changed the duration from seconds to milliseconds, in order to get the same time window, we had to adapt this value.
The reason: The test needs to pass in parallel, and with a high window interval (15s per test) this is not possible. Now, the duration of all tests has been reduced to 17ms.

File: src/engine/source/indexerconnector/test/unit/serverSelector_test.cpp (hunk 3 above)
Comment (GabrielEValenzuela) [reply src/engine/source/indexerconnector/test/unit/serverSelector_test.cpp:219]: How do we handle the yellow cluster? The previous tests only considered the green or red state. I think the logic is equal to the red server

Comment (sebasfalcone) [reply src/engine/source/indexerconnector/test/unit/serverSelector_test.cpp:219]: Yes, logic is equal to red server

Comment (GabrielEValenzuela) [reply src/engine/source/indexerconnector/test/unit/serverSelector_test.cpp:219]: Great, I will add a case for yellow server :)

Comment (GabrielEValenzuela) [reply src/engine/source/indexerconnector/test/unit/serverSelector_test.cpp:219]: Done!

File: src/engine/source/indexerconnector/src/monitoring.hpp (hunk 2 above)
Comment (GabrielEValenzuela) [reply src/engine/source/indexerconnector/src/monitoring.hpp]: Done!

File: src/engine/source/indexerconnector/test/unit/serverSelector_test.cpp [hunk 4]
```diff
@@ -129,31 +111,67 @@ TEST_F(ServerSelectorTest, TestGetNextBeforeHealthCheck)
...
+                std::string response;
+
+                if (url == GREEN_SERVER + "/_cat/health?v")
+                {
+                    // Simulate a successful response for the green server
+                    response = "epoch\ttimestamp\tcluster\tstatus\tnode.total\tnode.data\tdiscovered_cluster_"
+                               "manager\tshards\tpri\trelo\tinit\tunassign\tpending_tasks\tmax_task_wait_time\tactive_"
+                               "shards_percent\n1725296432\t17:00:32\twazuh-"
+                               "cluster\tgreen\t1\t1\ttrue\t47\t47\t0\t0\t0\t0\t-\t100.0%\n";
+                    postRequestParameters.onSuccess(response.c_str());
```
Comment (sebasfalcone) [inline src/engine/source/indexerconnector/test/unit/serverSelector_test.cpp:131]: We are repeating this responses in all tests, consider using a global variable inside a namespace

**Review by sebasfalcone — DISMISSED** [review]

File: src/engine/source/indexerconnector/test/unit/monitoring_test.cpp [hunk 5]
```diff
@@ -1,169 +1,154 @@
...
-        if (!m_fakeOpenSearchGreenServer)
+        const auto& url = requestParameters.url.url();
+        if (url == GREEN_SERVER + "/_cat/health?v")
         {
-            m_fakeOpenSearchGreenServer = std::make_unique<FakeOpenSearchServer>(host, 9209, "green");
+            postRequestParameters.onSuccess(greenResponse);
         }
-
-        if (!m_fakeOpenSearchRedServer)
+        else if (url == RED_SERVER + "/_cat/health?v" || url == YELLOW_SERVER + "/_cat/health?v")
```
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/engine/source/indexerconnector/test/unit/monitoring_test.cpp:28]: This was clarified in today's dayli, I believe it is not due to a lack of personal attention since before implementing this change I consulted you if the logic was the same and proceeded once you confirmed it (Ref: https://github.com/wazuh/wazuh/pull/25539#discussion_r1742844708).

Additionally, the logic implemented in `monitoring. hpp` currently only checks if the state is green, treating the other states similarly, and this is a unit test scenario.

In order to be able to close this task and avoid misunderstandings, the change request to be made is that the message to be sent is different if the server status is yellow or red?

File: src/engine/source/indexerconnector/test/unit/serverSelector_test.cpp (hunk 4 above)
Comment (GabrielEValenzuela) [reply src/engine/source/indexerconnector/test/unit/serverSelector_test.cpp:131]: I moved outside. I think a namespace it's not necessary, but let me know if you want, I add it

File: src/engine/source/indexerconnector/test/unit/monitoring_test.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (sebasfalcone) [reply src/engine/source/indexerconnector/test/unit/monitoring_test.cpp]: Miss communication happens, an improvement for the future 💪🏽

**Review by sebasfalcone — COMMENTED** [review]
LGTM!

File: src/engine/source/indexerconnector/test/unit/mockHTTTPRequest.hpp
Comment (Dwordcito) [inline src/engine/source/indexerconnector/test/unit/mockHTTTPRequest.hpp:1]: One thing, check the source folder structure.

Ex:
image

Comment (GabrielEValenzuela) [reply src/engine/source/indexerconnector/test/unit/mockHTTTPRequest.hpp:1]: Hi @Dwordcito ! 

That approach is different from the engine format, do you want to keep the current structure or change to the old version?

**Review by Dwordcito — APPROVED** [review]
